### DISCUSSION TITLE
update observable inside autorun/observable won't trigger render

### DISCUSSION THREAD

**Discussion by roeycohen** [mobxjs/mobx#2965] [Q&A] [answered]
Seems like updating an observable inside a rendering function will not behave nicely:

```
const testMap = observable.map({a: 1});
autorun(() =>
{
	console.log(testMap.toJS());
	if (!testMap.has('b'))
	{
		testMap.set('b', 2); //won't trigger autorun function again (and prevent following lines execution?)
		// setTimeout(() => testMap.set('b', 2), 0); //works
	}
});
```

Is it possible that this is by design to prevent infinite rendering?
If so, maybe this scenario is worth mentioning in the documentation pitfalls...

**Comment by BATCOH** [comment, marked as answer]
IDK about details of current implementation, but in general changing observables within an autorun\reaction is not recommended by the docs:
https://mobx.js.org/reactions.html#use-reactions-sparingly

**Reply by roeycohen** [reply]
1oq @BATCOH

**Comment by kubk** [comment]
Consider using a simple setter without any autorun / reactions:

```
...
setMap(key, value) {
  this.testMap.set(key, value);
  if (!testMap.has('b')) {
    this.testMap.set('b', 2);
  }
} 
```

**Reply by roeycohen** [reply]
but the problem @kubk occur when inside reactions, and not in a simple function :)

**Comment by urugator** [comment]
Possibly same as https://github.com/mobxjs/mobx/issues/2346, not entirely sure
